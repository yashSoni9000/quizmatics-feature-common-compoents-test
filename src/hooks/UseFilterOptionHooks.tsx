import { useEffect, useState } from "react";

export interface FilterListData {
    label: string;
    value: number;
}

export default function UseFilterOptionHooks(
    initialData: FilterListData[] = [],
    callBackFunction: () => void,
    getSearchList?: (search?: string) => void,
    initialSelectedValue?: number,
    initialIsLoading = true,
    initialIsError = false,
    initialErrorMessage?: string,
) {
    const [list, setList] = useState<FilterListData[]>(initialData);
    const [unfilteredList, setUnFilteredList] =
        useState<FilterListData[]>(initialData);
    const [searchText, setSearchText] = useState("");
    const [selectedValue, setSelectedValue] = useState<number | undefined>(
        initialSelectedValue,
    );
    const [isLoading, setIsLoading] = useState(initialIsLoading);
    const [error, setError] = useState(initialIsError);
    const [errorMessage, setErrorMessage] = useState(initialErrorMessage);

    useEffect(() => getSearchList && getSearchList(searchText), [searchText]);

    function updateSelectedValue(newSelectedValue: number) {
        if (selectedValue !== newSelectedValue)
            setSelectedValue(newSelectedValue);
    }

    function handleOnSearchChange(value?: string) {
        setSearchText(value || "");
    }

    function handleOnChange(value?: number) {
        if (value && value !== selectedValue) {
            updateSelectedValue(value);
        }
    }

    function refreshData() {
        callBackFunction();
    }

    function updateIsLoading(value: boolean) {
        setIsLoading(value);
    }

    function updateList(
        data: FilterListData[],
        error: boolean = false,
        message?: string,
        isLoading: boolean = false,
    ) {
        setList(data);
        setError(error);
        message && setErrorMessage(message);
        setIsLoading(isLoading);
    }

    function updateUnFilteredList(newList: FilterListData[]) {
        setUnFilteredList(newList);
    }

    return {
        list,
        isLoading,
        error,
        searchText,
        selectedValue,
        errorMessage,
        unfilteredList,
        updateList,
        updateIsLoading,
        updateSelectedValue,
        handleOnSearchChange,
        refreshData,
        handleOnChange,
        getSearchList,
        updateUnFilteredList,
    };
}
