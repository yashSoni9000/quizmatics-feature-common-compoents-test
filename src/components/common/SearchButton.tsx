import { FC, useState } from "react";
import {
    Box,
    IconButton,
    List,
    ListItem,
    ListItemButton,
    TextField,
    Popover,
    SvgIconProps,
} from "@mui/material";

import ErrorComponent from "@src/components/ErrorComponent";
import CircularLoader from "./CircularLoader";
import { EventFor } from "@src/ts/types";
import UseFilterOptionHooks from "@src/hooks/UseFilterOptionHooks";

export type SearchButtonProps = ReturnType<typeof UseFilterOptionHooks> & {
    Icon: FC<SvgIconProps>;
    placeholder?: string;
    updateData: () => void;
    size?: "small" | "medium";
};

export default function SearchButton({
    Icon,
    list,
    isLoading,
    error,
    handleOnChange,
    handleOnSearchChange,
    refreshData,
    getSearchList,
    placeholder,
    size = "small",
}: SearchButtonProps) {
    const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);
    const [searchText, setSearchText] = useState<string>("");

    const isDropdownOpen = Boolean(anchorEl);

    function setAnchorElement({
        currentTarget,
    }: EventFor<"button", "onClick">) {
        setAnchorEl(currentTarget);
        getSearchList && getSearchList();
    }

    function handleClose() {
        setAnchorEl(null);
    }

    function handleDropDownOnClick(value: number) {
        handleOnChange(value);
        handleClose();
    }

    function handleSearchFieldChange({
        currentTarget,
    }: EventFor<"input", "onChange">) {
        setSearchText(currentTarget.value);
        handleOnSearchChange(currentTarget.value);
    }

    return (
        <Box>
            <IconButton onClick={setAnchorElement} aria-label="Search Button">
                <Icon />
            </IconButton>
            <Popover
                className="rounded-lg"
                aria-label="Search Menu"
                open={isDropdownOpen}
                anchorEl={anchorEl}
                onClose={handleClose}
                anchorOrigin={{
                    vertical: "bottom",
                    horizontal: "left",
                }}
            >
                {isDropdownOpen && (
                    <Box
                        className="flex flex-col gap-2 min-w-[15rem] min-h-[15rem] h-fit max-h-[24rem] p-2"
                        sx={{ backgroundColor: "background.default" }}
                    >
                        <TextField
                            placeholder={placeholder}
                            value={searchText}
                            size={size}
                            onChange={handleSearchFieldChange}
                            autoFocus={true}
                            fullWidth
                            sx={{
                                backgroundColor: "background.textBoxColor",
                            }}
                        />
                        <ErrorComponent
                            error={error}
                            isLoading={isLoading}
                            refreshComponent={refreshData}
                        >
                            {isLoading ? (
                                <CircularLoader />
                            ) : (
                                <>
                                    {list.length !== 0 ? (
                                        <List>
                                            {list.map((item, index) => {
                                                const { value, label } = item;

                                                return (
                                                    <ListItem
                                                        key={index}
                                                        onClick={() =>
                                                            handleDropDownOnClick(
                                                                value,
                                                            )
                                                        }
                                                        value={value}
                                                        className="w-full p-0"
                                                    >
                                                        <ListItemButton>
                                                            {label}
                                                        </ListItemButton>
                                                    </ListItem>
                                                );
                                            })}
                                        </List>
                                    ) : (
                                        <Box className="mt-2 p-2">
                                            No results found
                                        </Box>
                                    )}
                                </>
                            )}
                        </ErrorComponent>
                    </Box>
                )}
            </Popover>
        </Box>
    );
}
