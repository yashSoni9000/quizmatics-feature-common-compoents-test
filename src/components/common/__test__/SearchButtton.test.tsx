import {
    fireEvent,
    render,
    renderHook,
    screen,
    within,
} from "@testing-library/react";
import "@testing-library/jest-dom";

import SearchButton, { SearchButtonProps } from "../SearchButton";
import { act } from "react-dom/test-utils";

import UseFilterOptionHooks, {
    FilterListData,
} from "@src/hooks/UseFilterOptionHooks";

let updateDataMock: ReturnType<typeof jest.fn>;
let callbackMock: ReturnType<typeof jest.fn>;
let onChangeMock: ReturnType<typeof jest.fn>;

let filterHook: ReturnType<typeof UseFilterOptionHooks>;
const mockData = [
    { label: "Option 1", value: 1 },
    { label: "Option 2", value: 2 },
];

beforeEach(() => {
    updateDataMock = jest.fn();
    callbackMock = jest.fn();
    onChangeMock = jest.fn();

    filterHook = renderHook(() =>
        UseFilterOptionHooks(
            mockData,
            callbackMock,
            jest.fn(),
            undefined,
            false,
        ),
    ).result.current;
});

function SearchButtonComponent({ ...props }: Partial<SearchButtonProps>) {
    return (
        <SearchButton
            {...filterHook}
            Icon={jest.fn()}
            updateData={updateDataMock}
            {...props}
        />
    );
}

describe("Search Button Test", () => {
    it("snapshot test", () => {
        const { asFragment } = render(<SearchButtonComponent />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("Clicking on search button opens dropdown", () => {
        render(<SearchButtonComponent />);

        const searchButton = screen.getByLabelText("Search Button");
        fireEvent.click(searchButton);

        const searchMenu = screen.getByLabelText("Search Menu");
        expect(searchMenu).toBeVisible();
    });

    it("Clicking on menu item invokes on click and hides the menu", async () => {
        render(<SearchButtonComponent handleOnChange={onChangeMock} />);

        const searchButton = screen.getByLabelText("Search Button");
        fireEvent.click(searchButton);

        const listItems = screen.getAllByRole("listitem");
        expect(listItems).toHaveLength(filterHook.list.length);

        const listItem = screen.getByText(mockData[0].label);
        await act(() => fireEvent.click(listItem));

        expect(onChangeMock).toHaveBeenCalledWith(mockData[0].value);

        const searchMenu = screen.getByLabelText("Search Menu");
        expect(searchMenu).toHaveAttribute("aria-hidden", "true");
    });

    it("Typing in search Field changes the value and hides the unmatched options", () => {
        let filteredData: FilterListData[] = mockData;

        function handleOnSearchChangeMock(value?: string) {
            if (value) {
                filteredData = mockData.filter(({ label }) =>
                    label.includes(value),
                );
            } else {
                filteredData = mockData;
            }
        }

        function GenericSearchButtonComponent() {
            return (
                <SearchButtonComponent
                    list={filteredData}
                    handleOnSearchChange={handleOnSearchChangeMock}
                />
            );
        }

        function rerenderComponent() {
            rerender(<GenericSearchButtonComponent />);
        }

        const { rerender } = render(<GenericSearchButtonComponent />);

        const searchButton = screen.getByLabelText("Search Button");
        fireEvent.click(searchButton);

        const searchField = screen.getByRole("textbox");
        const list = screen.getByRole("list");

        fireEvent.change(searchField, { target: { value: mockData[0].label } });

        rerenderComponent();

        const optionOne = within(list).queryByText(mockData[0].label);
        const optionTwo = within(list).queryByText(mockData[1].label);

        expect(optionOne).toBeInTheDocument();
        expect(optionTwo).not.toBeInTheDocument();
    });
});
