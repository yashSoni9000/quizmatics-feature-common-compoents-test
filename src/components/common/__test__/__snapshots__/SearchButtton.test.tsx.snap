// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Button Test snapshot test 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <button
      aria-label="Search Button"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
